### PR TITLE
High Level Streams

### PR DESCRIPTION
**Pull request by amandaratio**
1. initiates stream for `PUBLISH`
2. initiates stream for `SUBSCRIBE`
3. Handles stream for `SUBSCRIBE` with proceeding `PUBLISH`

### CODE REVIEW

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 1]
```diff
@@ -275,7 +281,8 @@ private RouteFW resolveTarget(
         private DecoderState decodeState;
         private int slotIndex = NO_SLOT;
         private int slotLimit;
-        private HashMap<String, MqttServerStream> mqttStreams;
+        private HashMap<String, MqttServerStream> mqttSubscriptionStreams;
+        private HashMap<String, MqttServerStream> mqttPublishStreams;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java]: Please change these types to `private final Map<String, MqttServerStream>` instead.
Suggest renaming the fields to `subscribers` and `publishers`.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 2]
```diff
@@ -586,6 +594,61 @@ private void onMqttUnsubscribe(
         private void onMqttPublish(
             MqttPublishFW publish)
         {
+            if (publish != null)
+            {
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java]: Do we need to test for `null` here or is `publish` guaranteed to always be `non-null`?

**Review by jfallows — CHANGES_REQUESTED** [review]

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 3]
```diff
@@ -275,7 +284,8 @@ private RouteFW resolveTarget(
         private DecoderState decodeState;
         private int slotIndex = NO_SLOT;
         private int slotLimit;
-        private HashMap<String, MqttServerStream> mqttStreams;
+        private Map<String, MqttServerStream> subscribers;
+        private Map<String, MqttServerStream> publishers;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:288]: Please mark these as `final` and move up with other `final` fields.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 4]
```diff
@@ -872,19 +955,87 @@ private void onApplication(
...
+            final WindowFW window = windowRW.wrap(writeBuffer, 0, writeBuffer.capacity())
+                .routeId(routeId)
+                .streamId(replyId)
+                .trace(supplyTraceId.getAsLong())
+                .credit(bufferPool.slotCapacity())
+                .padding(0)
+                .groupId(0)
+                .build();
 
+            application.accept(window.typeId(), window.buffer(), window.offset(), window.sizeof());
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:967]: Recommend hoisting the above logic into `doWindow` method on `MqttServer`.

When we receive the `BEGIN` reply here, it should trigger the sending of `SUBACK` frame if the initial `MqttServerStream` was created for `SUBSCRIBE`, right?

So that would be `server.onSubscribed(...)` and initially we can assume just one topic filter per `SUBSCRIBE`.

Next step would be to defer sending the `SUBACK` frame until all `reason` codes are collected from multiple application streams, triggered by sending parallel application `BEGIN` streams in reaction to the same inbound `SUBSCRIBE` frame.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 5]
```diff
@@ -872,19 +955,87 @@ private void onApplication(
...
+        }
+
+        private void onData(
+            DataFW data)
+        {
+            OctetsFW ex = data.extension();
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:973]: `final` and rename to `extension`.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 6]
```diff
@@ -872,19 +955,87 @@ private void onApplication(
...
+
+            final DirectBuffer exBuffer = data.extension().buffer();
+            final int exOffset = data.extension().offset();
+            final int exLength = data.extension().sizeof();
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:977]: Use `extension` variable instead of calling `data.extension()` multiple times.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 7]
```diff
@@ -842,8 +901,24 @@ private int decodeSession(
...
+        private long replyId;
+
+        MqttServerStream(
+            MqttServer network,
+            MessageConsumer application,
+            long routeId,
+            long initialId,
+            long replyId)
         {
+            this.receiver = network;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:917]: Rename to `this.server = server`.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 8]
```diff
@@ -872,19 +955,87 @@ private void onApplication(
...
+            final DataFW dataFW = dataRW.wrap(writeBuffer, 0, writeBuffer.capacity())
+                .routeId(receiver.routeId)
+                .streamId(receiver.replyId)
+                .trace(supplyTraceId.getAsLong())
+                .groupId(0)
+                .padding(0)
+                .payload(publishBuffer, publishOffset, publishLength)
+                .build();
+
+            this.receiver.network.accept(dataFW.typeId(), dataFW.buffer(), dataFW.offset(), dataFW.sizeof());
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:1008]: This becomes `server.network.accept(...)`

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java [hunk 9]
```diff
@@ -872,19 +955,87 @@ private void onApplication(
...
         private void onReset(
             ResetFW reset)
         {
+        }
+
+        private void doBegin(
+            MessageConsumer target,
+            long routeId,
+            long streamId,
+            long traceId)
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:1030]: Most of these arguments are already available as member variables and can be removed.
Please apply same feedback to other `doXXX` methods on `MqttServerStream`.

File: src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java (hunk 8 above)
Comment (jfallows) [reply src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java:1008]: Also refactor above code from `983` onwards into `server.doMqttPublish(...)`

**Review by jfallows — CHANGES_REQUESTED** [review]
